File: tst/triggers/selling/sales_order/sales_order.py
import frappe
from frappe.model.mapper import get_mapped_doc
from frappe import _


@frappe.whitelist()
def before_validate(doc, method=None):
    set_contract_type(doc)


@frappe.whitelist()
def make_training(sales_order_name):
    """
    Create Training documents for all Device Setups linked to this Sales Order
    Returns the last created Training name for routing
    """
    if not sales_order_name:
        frappe.throw(_("Sales Order name is required"))

    # Get all submitted Device Setups for this Sales Order
    device_setups = frappe.get_all(
        "Device Setup",
        filters={"sales_order": sales_order_name, "docstatus": 1},
        fields=["name", "serial_no"],
    )

    if not device_setups:
        frappe.throw(_("No submitted Device Setups found for this Sales Order"))

    last_training = None

    for device in device_setups:
        # Check if Training already exists
        # if not frappe.db.exists("Training", {"device": device.name}):
        training = frappe.new_doc("Training")
        training.update(
            {
                "device": device.name,
                "serial_no": device.serial_no,  # Include serial no if needed
                "sales_order": sales_order_name,
                "account_setup_verified": 1,
                "posting_date": frappe.utils.nowdate(),
            }
        )
        training.insert(ignore_permissions=True)
        last_training = training.name

    if not last_training:
        frappe.msgprint(_("Trainings already exist for all Device Setups"))
        return None

    return last_training


@frappe.whitelist()
def make_installation_order(source_name, target_doc=None):
    args = frappe.form_dict.get("args")
    if isinstance(args, str):
        args = frappe.parse_json(args)

    address = args.get("address")
    parent_address = args.get("parent_address")

    if not address:
        frappe.throw("Delivery address is required to create the Installation Order.")

    def item_filter(doc):
        return True if parent_address else doc.custom_address == address

    def postprocess(source_doc, target_doc):
        target_doc.customer_address = address

    doc = get_mapped_doc(
        "Sales Order",
        source_name,
        {
            "Sales Order": {
                "doctype": "Installation Order",
                "field_map": {
                    "customer": "customer",
                    "name": "sales_order",
                    "delivery_date": "installation_date",
                },
            },
            "Sales Order Item": {
                "doctype": "Installation Order Item",
                "field_map": {"custom_contact": "contact"},
                "condition": item_filter,
            },
        },
        target_doc,
        postprocess=postprocess,
    )
    customer_email = frappe.db.get_value(
        "Customer",
        frappe.db.get_value("Sales Order", source_name, "customer"),
        "email_id",
    )
    doc.customer_email = customer_email

    return doc


def set_contract_type(doc):
    """
    Set the custom_contract_type for the current Sales Order based on the total custom_no_of_cars
    for the same customer across all their Sales Orders.
    """
    if not doc.customer:
        frappe.throw(_("Customer is required to set the contract type."))

    # Calculate the total custom_no_of_cars for this customer across all their sales orders
    total_cars_for_customer = frappe.db.sum(
        "Sales Order",
        filters={"customer": doc.customer, "docstatus": 1},  # Only consider submitted orders
        fieldname="custom_no_of_cars",
    )

    # Determine the contract type based on the total cars for this customer
    if total_cars_for_customer and total_cars_for_customer >= 5:
        doc.custom_contract_type = "B2B"
    else:
        doc.custom_contract_type = "B2C"


@frappe.whitelist()
def get_customer_address_and_contacts_list(customer):
    if not customer:
        return []

    addresses = frappe.get_all(
        "Dynamic Link",
        filters={
            "link_doctype": "Customer",
            "link_name": customer,
            "parenttype": "Address",
        },
        pluck="parent",
    )

    contacts = frappe.get_all(
        "Dynamic Link",
        filters={
            "link_doctype": "Customer",
            "link_name": customer,
            "parenttype": "Contact",
        },
        pluck="parent",
    )

    return addresses, contacts

File: tst/triggers/selling/sales_order/sales_order.js
frappe.ui.form.on('Sales Order', {
    onload_post_render: function(frm) {
            cur_frm.remove_custom_button(__('Update Items'));
            cur_frm.remove_custom_button(__('Status'));
            cur_frm.remove_custom_button(__('Pick List'), __('Create'));
            // cur_frm.remove_custom_button(__('Delivery Note'), __('Create'));
            cur_frm.remove_custom_button(__('Work Order'), __('Create'));
            cur_frm.remove_custom_button(__('Request for Raw Materials'), __('Create'));
            cur_frm.remove_custom_button(__('Project'), __('Create'));
            cur_frm.remove_custom_button(__('Payment Request'), __('Create'));
            cur_frm.remove_custom_button(__('Purchase Order'), __('Create'));
            let statusButton = $(`[data-label="${__('Status')}"]`);
            statusButton.hide();
    },
    refresh:  function (frm) {
        cur_frm.remove_custom_button(__('Update Items'));
            cur_frm.remove_custom_button(__('Status'));
            cur_frm.remove_custom_button(__('Pick List'), __('Create'));
            // cur_frm.remove_custom_button(__('Delivery Note'), __('Create'));
            cur_frm.remove_custom_button(__('Work Order'), __('Create'));
            cur_frm.remove_custom_button(__('Request for Raw Materials'), __('Create'));
            cur_frm.remove_custom_button(__('Project'), __('Create'));
            cur_frm.remove_custom_button(__('Payment Request'), __('Create'));
            cur_frm.remove_custom_button(__('Purchase Order'), __('Create'));
            let statusButton = $(`[data-label="${__('Status')}"]`);
            statusButton.hide();
        address_filter(frm)
        if (frm.doc.docstatus == 1 ) {
            frm.add_custom_button(__('Installation Order'), function () {

                // Step 1: Get unique addresses from items
                let address_list = [];
                let parent_address = false;
                (frm.doc.items || []).forEach(item => {
                    if (item.custom_address && !address_list.includes(item.custom_address)) {
                        address_list.push(item.custom_address);
                    }
                });

                // Step 2: Handle based on number of addresses
                if (address_list.length === 0) {
                    address_list.push(frm.doc.customer_address);
                    parent_address = true;

                    // frappe.msgprint(__('No delivery addresses found in the items.'));
                    // return;
                }

                if (address_list.length === 1) {
                    // Only one address, proceed
                    create_installation_order(frm, address_list[0],parent_address);
                } else {
                    // Multiple addresses, show dialog
                    let dialog = new frappe.ui.Dialog({
                        title: 'Select Delivery Address',
                        fields: [
                            {
                                fieldname: 'selected_address',
                                label: 'Delivery Address',
                                fieldtype: 'Select',
                                options: address_list,
                                reqd: 1
                            }
                        ],
                        primary_action_label: 'Create Installation Order',
                        primary_action(values) {
                            dialog.hide();
                            create_installation_order(frm, values.selected_address,parent_address);
                        }
                    });
                    dialog.show();
                }

            });
            frappe.call({
            method: 'frappe.client.get_value',
            args: {
                doctype: 'Device Setup',
                filters: { sales_order: frm.doc.name },
                fieldname: 'name'
            },
            callback: function(r) {
                
                if (!r.exc && r.message && r.message.name) {
                    // Prevent duplicate button
                    if (!frm.custom_buttons || !frm.custom_buttons['Start Training']) {
                        frm.add_custom_button(__('Start Training'), () => {
                            frappe.call({
                                method: 'tst.triggers.selling.sales_order.sales_order.make_training',
                                args: {
                                    sales_order_name: frm.doc.name
                                },
                                callback: function(response) {
                                    if (response.message) {
                                        frappe.set_route('Form', 'Training');
                                    }
                                },
                                freeze: true,
                                freeze_message: __('Creating Training...')
                            });
                        }).addClass('btn-primary');
                    }
                }
            }
        });
    }
    }
});

function address_filter(frm) {
    if (frm.doc.customer){
        frappe.call({
            method: 'tst.triggers.selling.sales_order.sales_order.get_customer_address_and_contacts_list',
            args: {
                customer: frm.doc.customer
            },
            callback: function(response) {
                // console.log(response.message)
                const address_list = response.message[0] || [];
                const contact_list = response.message[1] || [];

                frm.fields_dict['items'].grid.get_field("custom_address").get_query = function(doc, cdt, cdn) {
                    return {
                        filters: [
                            ["Address", "name", "in", address_list]
                        ]
                    };
                };
                frm.fields_dict['items'].grid.get_field("custom_contact").get_query = function(doc, cdt, cdn) {
                    return {
                        filters: [
                            ["Contact", "name", "in", contact_list]
                        ]
                    };
                };
                
            }
        });
    }
}

// Function to call your Python method with selected address
function create_installation_order(frm, address,parent_address) {
    frappe.model.open_mapped_doc({
        method: 'tst.triggers.selling.sales_order.sales_order.make_installation_order',
        frm: frm,
        args: {
            address: address , // Pass the selected address to server-side method
            parent_address:parent_address
        }
    });
}


// frappe.ui.form.on("Sales Order Item",{
//     custom_address:function(frm){
//         console.log("testststss")
//         if (frm.doc.customer){
//             address_filter(frm)
//         }
//     },
    
// })
